File: core/domain/src/testFixtures/java/com/github/mcgalanes/groomr/core/domain/fixture/DomainFixtures.java
package com.github.mcgalanes.groomr.core.domain.fixture;

import com.github.mcgalanes.groomr.core.domain.model.UserStory;
import com.github.mcgalanes.groomr.core.domain.model.UserStory.Criteria;
import com.github.mcgalanes.groomr.core.domain.model.UserStory.Criteria.GherkinLine;
import com.github.mcgalanes.groomr.core.domain.model.UserStory.Criteria.GherkinLine.GherkinKey;

import java.util.ArrayList;
import java.util.List;

import kotlin.random.Random;


public class DomainFixtures {
    private static final Random random = Random.Default;

    public static UserStory randomUserStory() {
        List<Criteria> criteriaList = new ArrayList<>();
        criteriaList.add(randomCriteria());
        criteriaList.add(randomCriteria());
        criteriaList.add(randomCriteria());

        return new UserStory(
            random.nextLong(1, Long.MAX_VALUE),
            "User Story " + random.nextInt(1, 100),
            "Persona " + random.nextInt(1, 100),
            "Wish " + random.nextInt(1, 100),
            "Purpose " + random.nextInt(1, 100),
            "KPI " + random.nextInt(1, 100),
            random.nextInt(1, 100),
            "Solution " + random.nextInt(1, 100),
            "Enablers " + random.nextInt(1, 100),
            "Assets " + random.nextInt(1, 100),
            random.nextInt(1, 100),
            random.nextBoolean(),
            random.nextBoolean(),
            random.nextBoolean(),
            random.nextBoolean(),
            criteriaList
        );
    }

    public static Criteria randomCriteria() {
        List<GherkinLine> gherkinLines = new ArrayList<>();
        gherkinLines.add(randomGherkinLine());
        gherkinLines.add(randomGherkinLine());
        gherkinLines.add(randomGherkinLine());

        return new Criteria(
            random.nextLong(1, Long.MAX_VALUE),
            "Title " + random.nextInt(1, 100),
            gherkinLines
        );
    }

    public static GherkinLine randomGherkinLine() {
        Random random = Random.Default;

        GherkinKey[] keys = GherkinKey
            .getEntries()
            .toArray(GherkinKey[]::new);

        return new GherkinLine(
            random.nextLong(1, Long.MAX_VALUE),
            keys[random.nextInt(keys.length)],
            "Value " + random.nextInt(1, 100)
        );
    }
}

File: core/data/src/test/java/com/github/mcgalanes/groomr/core/data/fixture/DomainFixtures.kt
package com.github.mcgalanes.groomr.core.data.fixture

import com.github.mcgalanes.groomr.core.domain.model.UserStory
import kotlin.random.Random
import kotlin.random.nextInt
import kotlin.random.nextLong

fun Random.nextUserStory(
    criteriaList: List<UserStory.Criteria> = listOf(
        Random.nextCriteria(),
        Random.nextCriteria(),
        Random.nextCriteria(),
    ),
): UserStory =
    UserStory(
        id = nextLong(1..Long.MAX_VALUE),
        title = "User Story ${nextInt(1..100)}",
        persona = "Persona ${nextInt(1..100)}",
        wish = "Wish ${nextInt(1..100)}",
        purpose = "Purpose ${nextInt(1..100)}",
        kpi = "KPI ${nextInt(1..100)}",
        businessValue = nextInt(1..100),
        solution = "Solution ${nextInt(1..100)}",
        enablers = "Enablers ${nextInt(1..100)}",
        assets = "Assets ${nextInt(1..100)}",
        estimation = nextInt(1..100),
        smallEnough = nextBoolean(),
        independent = nextBoolean(),
        estimable = nextBoolean(),
        testable = nextBoolean(),
        criteriaList = criteriaList,
    )

fun Random.nextCriteria(): UserStory.Criteria =
    UserStory.Criteria(
        title = "Title ${nextInt(1..100)}",
        gherkinLines = listOf(
            Random.nextGherkinLine(),
            Random.nextGherkinLine(),
            Random.nextGherkinLine(),
        ),
    )

fun Random.nextGherkinLine(): UserStory.Criteria.GherkinLine =
    UserStory.Criteria.GherkinLine(
        key = UserStory.Criteria.GherkinLine.GherkinKey.entries.toTypedArray().random(),
        value = "Value ${nextInt(1..100)}",
    )
